erasure-readfile.go

```diff
@@ -17,54 +17,47 @@
 package main
 
 import (
+	"bytes"
 	"encoding/hex"
 	"errors"
+	"io"
 
 	"github.com/klauspost/reedsolomon"
 )
 
-// erasureReadFile - read an entire erasure coded file at into a byte
-// array. Erasure coded parts are often few mega bytes in size and it
-// is convenient to return them as byte slice. This function also
-// supports bit-rot detection by verifying checksum of individual
-// block's checksum.
-func erasureReadFile(disks []StorageAPI, volume string, path string, partName string, size int64, eInfos []erasureInfo) ([]byte, error) {
-	// Return data buffer.
-	var buffer []byte
-
-	// Total size left
-	totalSizeLeft := size
-
-	// Starting offset for reading.
-	startOffset := int64(0)
+// erasureReadFile - read bytes from erasure coded files and writes to given writer.
+// Erasure coded files are read block by block as per given erasureInfo and shards
+// are decoded into a data block.  Data block is trimmed for given offset and length,
+// then written to given writer.
+func erasureReadFile(writer io.Writer, disks []StorageAPI, volume string, path string, partName string, eInfos []erasureInfo, offset int64, length int64) (int64, error) {
+	// Total bytes written to writer
+	bytesWritten := int64(0)
 
 	// Gather previously calculated block checksums.
 	blockCheckSums := metaPartBlockChecksums(disks, eInfos, partName)
 
 	// Pick one erasure info.
 	eInfo := pickValidErasureInfo(eInfos)
 
-	// Write until each parts are read and exhausted.
-	for totalSizeLeft > 0 {
-		// Calculate the proper block size.
-		var curBlockSize int64
-		if eInfo.BlockSize < totalSizeLeft {
-			curBlockSize = eInfo.BlockSize
-		} else {
-			curBlockSize = totalSizeLeft
-		}
+	// Calculate start block for given offset and how many bytes to skip to get the offset.
+	startBlock := offset / eInfo.BlockSize
```

Comment:
Are you talking about moving start/end block calculation into a separate function?